proposals/0002-monitoring.md

```diff
@@ -0,0 +1,178 @@
+
+Data access must honour the http response codes, especially
+
+* 200 for successful retrieval of an object
+* 204 when retrieving a subtree that would exist, but has no content. E.g. when the application-specific subtree has no application specific metrics defined.
+* 404 if an directly-addressed item does not exist. This may be a non-existing sub-tree or non-existing object
+* 503 to indicate that a health check resulted in "bad health". The body SHOULD contain details { "details": <text> }
+
+The Api MUST NOT return a 500 Internal Server Error code to represent a non-existing resource or to indicate that a health-check resulted in "bad health"
+
+## Detailed design (if applicable)
+
+### Api Objects
+
+Api-objects MAY include one or more metrics as in 
+
+
+    {
+      "thread-count" : 33,
+      "peak-thread-count" : 47,
+      "total-started-thread-count" : 49,
+      "current-thread-cpu-time" : 66855000,
+      "current-thread-user-time" : 64003000
+    }
+
+or
+
+    {
+      "count": 45
+    }  
+
+### Metadata
+
+Metric values MUST have metadata 
+
+* unit: a fixed set of string units from e.g. [1], [UoM] or [Metrics2.0]  
+* type: 
+     * counter: a monotonically increasing or decreasing numeric value (e.g. total number of requests received)
+     * gauge: a numeric value that can arbitrarily go up and down
+     * bool: a boolean value which can be `true` or `false`
+     * string: a string
+* description (optional): A human readable description of the metric
+* displayName (optional): A human readable name of the metric for display purposes if the metric name is not
+human readable. This could e.g. be the case when the metric name is a uuid.
+* tags (optional): A list of key=value pairs, which are separated by comma. 
+
+Example: 
+
+if `GET /metrics/foo` exposes:
+
+      {"fooVal": 12345} 
+
+then `OPTIONS /metrics/foo` will expose:
+
+    {"fooVal": {
+          "unit": "ms",
+          "type": "gauge",
+          "description": "The size of foo after each request",
+          "displayName": "Size of foo",
+          "tags": "app=webshop
+         }
+    } 
+
+Metadata must not change over the lifetime of a process (i.e. it is not allowed
+to return the units as seconds in one retrieval and as hours in a subsequent one).
+The reason behind it is that e.g. a monitoring agent on Kubernetes may read the
+metadata once it sees the new container and store it. It may not periodically
+re-query the process for the metadata.
+
+In fact that metadata should probably not change during the life-time of the
+whole container image or an application, as all containers spawned from it
+will be "the same" and form part of an app, where it would be confusing in
+an overall view if the same metric has different metadata.
+
+Metadata SHOULD support caching via cache control headers and SHOULD reply with a 304 Not Modified response accordingly.
+
+## Required metrics
+
+Required metrics are exposed under `/metrics/base`
```

Comment:
Can we have a small description in this section saying required metrics are the metrics that all vendors must implement? I could only understand it reading the vendor specific data section.

Also, can we start describing which metrics will be required?
- Memory
- Disk
- CPU
- JVM
- Database pool
- Network
- Processes